hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/OnDemandContainerScanner.java

```diff
@@ -0,0 +1,165 @@
+    if (addContainerToScheduledContainers(containerId)) {
+      resultFuture = instance.scanExecutor.submit(() -> {
+        removeContainerFromScheduledContainers(containerId);
+        if (container.shouldScanData()) {
+          performOnDemandScan(container);
+        }
+      });
+    }
+    return Optional.ofNullable(resultFuture);
+  }
+
+  private static boolean addContainerToScheduledContainers(long containerId) {
+    return instance.containerRescheduleCheckSet.add(containerId);
+  }
+
+  private static void removeContainerFromScheduledContainers(
+      long containerId) {
+    instance.containerRescheduleCheckSet.remove(containerId);
+  }
+
+  private static void performOnDemandScan(Container<?> container) {
+    long containerId = container.getContainerData().getContainerID();
+    try {
+      ContainerData containerData = container.getContainerData();
+      logScanStart(containerData);
+      if (container.scanData(instance.throttler, instance.canceler)) {
+        Instant now = Instant.now();
+        logScanCompleted(containerData, now);
+        instance.containerController.updateDataScanTimestamp(containerId, now);
+      } else {
+        instance.containerController.markContainerUnhealthy(containerId);
+        instance.metrics.incNumUnHealthyContainers();
+      }
+      instance.metrics.incNumContainersScanned();
+    } catch (IOException e) {
+      LOG.warn("Unexpected exception while scanning container "
+          + containerId, e);
+    }
+  }
+
+  private static void logScanStart(ContainerData containerData) {
+    if (LOG.isDebugEnabled()) {
+      Optional<Instant> scanTimestamp = containerData.lastDataScanTime();
+      Object lastScanTime = scanTimestamp.map(ts -> "at " + ts).orElse("never");
+      LOG.debug("Scanning container {}, last scanned {}",
+          containerData.getContainerID(), lastScanTime);
+    }
+  }
+
+  private static void logScanCompleted(
+      ContainerData containerData, Instant timestamp) {
+    LOG.debug("Completed scan of container {} at {}",
+        containerData.getContainerID(), timestamp);
+  }
+
+  public static OnDemandScannerMetrics getMetrics() {
+    return instance.metrics;
+  }
+
+  public static synchronized void shutdown() {
+    if (instance == null) {
+      return;
+    }
+    instance.shutdownScanner();
+  }
+
+  private synchronized void shutdownScanner() {
+    instance = null;
+    metrics.unregister();
+    if (!scanExecutor.isShutdown()) {
+      scanExecutor.shutdown();
+    }
+    try {
+      long timeoutSeconds = 5;
+      if (!scanExecutor.awaitTermination(timeoutSeconds, TimeUnit.SECONDS)) {
+        LOG.warn("On demand scanner shut down forcefully after {} seconds",
+            timeoutSeconds);
+        scanExecutor.shutdownNow();
+      }
+    } catch (InterruptedException e) {
+      LOG.warn("On demand scanner interrupted while waiting for shut down.");
+      scanExecutor.shutdownNow();
+      throw new RuntimeException(e);
```

Comment:
I think we should set the interrupt flag and return instead of throwing. See `AbstractContainerScanner#shutdown` for reference.